drivers/adc/ad7091r8/ad7091r8.c

```diff
@@ -0,0 +1,558 @@
+	int ret;
+
+	if (!dev || !reg_data)
+		return -EINVAL;
+
+	if (reg_addr == AD7091R8_REG_RESULT) {
+		ret = ad7091r8_pulse_convst(dev);
+		if (ret)
+			return ret;
+	}
+
+	no_os_put_unaligned_be16(
+		no_os_field_prep(AD7091R8_RD_WR_FLAG_MSK, 0) |
+		no_os_field_prep(AD7091R8_REG_ADDR_MSK, reg_addr), buf);
+
+	/* Reg data only comes out on the next transfer (datasheet figure 52) */
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, 2);
+	if (ret)
+		return ret;
+
+	if (reg_addr == AD7091R8_REG_RESULT) {
+		ret = ad7091r8_pulse_convst(dev);
+		if (ret)
+			return ret;
+	}
+
+	no_os_put_unaligned_be16(
+		no_os_field_prep(AD7091R8_RD_WR_FLAG_MSK, 0) |
+		no_os_field_prep(AD7091R8_REG_ADDR_MSK, reg_addr), buf);
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, 2);
+	if (ret)
+		return ret;
+
+	*reg_data = no_os_get_unaligned_be16(buf);
+	return 0;
+}
+
+/**
+ * SPI write to device using a mask.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - The mask.
+ * @param data - The register data.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_spi_write_mask(struct ad7091r8_dev *dev,
+			    uint8_t reg_addr,
+			    uint16_t mask,
+			    uint16_t data)
+{
+	uint16_t reg_data;
+	int ret;
+
+	ret = ad7091r8_spi_reg_read(dev, reg_addr, &reg_data);
+	if (ret)
+		return ret;
+
+	reg_data &= ~mask;
+	reg_data |= data;
+
+	return ad7091r8_spi_reg_write(dev, reg_addr, reg_data);
+}
+
+/**
+ * @brief Set device sleep mode.
+ *
+ * @param dev - The device structure.
+ * @param mode - The device sleep mode to set.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_set_sleep_mode(struct ad7091r8_dev *dev,
+			    enum ad7091r8_sleep_mode mode)
+{
+	return ad7091r8_spi_write_mask(dev, AD7091R8_REG_CONF,
+				       REG_CONF_SLEEP_MODE_MASK, mode);
+}
+
+/**
+ * Set output port value.
+ * @param dev - The device structure.
+ * @param port - Port number.
+ * @param value - Value.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_set_port(struct ad7091r8_dev *dev, enum ad7091r8_port port,
+		      bool value)
+{
+	uint16_t mask;
+	uint16_t val;
+
+	switch (port) {
+	case AD7091R8_GPO0:
+		mask = REG_CONF_GPO0_MASK;
+		val = no_os_field_get(REG_CONF_GPO0_MASK, value);
+		break;
+	case AD7091R8_GPO1:
+		mask = REG_CONF_GPO1_MASK;
+		val = no_os_field_get(REG_CONF_GPO1_MASK, value);
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7091r8_spi_write_mask(dev, AD7091R8_REG_CONF, mask, val);
+}
+
+/**
+ * Set GPO0 mode.
+ * @param dev - The device structure.
+ * @param mode - GPO0 new mode.
+ * @param is_cmos - 0: GPO0 is open drain
+ * 		  - 1: GPO0 is CMOS.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int ad7091r8_set_gpo0_mode(struct ad7091r8_dev *dev,
+			   enum ad7091r8_gpo0_mode mode, bool is_cmos)
+{
+	uint16_t value;
+
+	switch (mode) {
+	case AD7091R8_GPO0_ENABLED:
```

Comment:
hmm, what's the point of doing this switch case when you have the `enum ad7091r8_gpo0_mode mode` and you can initialized the reg values directly there:
```
/**
 * @enum ad7091r8_gpo0_mode
 * @brief Port 0 configuration
 */
enum ad7091r8_gpo0_mode {
	/** GPO0 is output port */
	AD7091R8_GPO0_ENABLED = 0,
	/** GPO0 is Alert indicator */
	AD7091R8_GPO0_ALERT = 16,
	/** GPO0 is busy indicator, device is converting */
	AD7091R8_GPO0_BUSY = 48,
};
```

Reply:
Nice. thanks!